metrics_api/lib/opentelemetry/metrics/instrument/asynchronous_instrument.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Metrics
+    module Instrument
+      # https://opentelemetry.io/docs/reference/specification/metrics/api/#asynchronous-instrument-api
+      class AsynchronousInstrument
+        attr_reader :name, :unit, :description, :callback
+
+        # @api private
+        def initialize(name, unit: nil, description: nil, callback: nil)
+          @name = name
+          @unit = unit || ''
+          @description = description || ''
+          @callback = callback ? Array(callback) : []
+        end
+
+        # @param callback [Proc, Array<Proc>]
+        #   Callback functions should:
+        #   - be reentrant safe;
+        #   - not take an indefinite amount of time;
+        #   - not make duplicate observations (more than one Measurement with the same attributes)
+        #     across all registered callbacks;
+        def register_callback(callback); end
```

Comment:
Similarly, the plural form is preferable. I'd also like to use the "splat" form: 
```suggestion
        def register_callbacks(*callback); end
```
This allows calling as:
```ruby
instr.register_callbacks(cb1)
instr.register_callbacks(cb2, cb3)
```
instead of
```ruby
instr.register_callbacks(cb1)
instr.register_callbacks([cb2, cb3])
```

Reply:
I like the splat form, but I've been thinking if should actually support all those cases later in the implementation:

```rb
instr.register_callback(cb1)
instr.register_callback(cb2, cb3)
instr.register_callback([cb4, cb5]) # And support this as well?
# Just so it is consistent with `callbacks:` from the `Meter#create_observable_counter` API
```

In the implementation, I think we could support that with

```rb
def register_callbacks(*callbacks)
  @callbacks.concat(callbacks.flatten)
end
```

Because otherwise it would feel odd to be able to pass an array of procs when creating an instrument, like

```rb
meter = MeterProvider.get_meter('test')
instr = meter.create_observable_counter('test', callbacks: [cb1, cb2])
```

but not be able to pass an array of procs when registering callbacks later

```rb
meter = MeterProvider.get_meter('test')
instr = meter.create_observable_counter('test')

instr.register_callbacks([cb1, cb2]) # ?
```